### ISSUE TITLE
More comprehensive rustdocs

### ISSUE THREAD

**Issue by hniksic** [andrewhickman/fs-err#45]
The docs of the `fs-err` wrappers currently just say "Wrapper for <function>". While that's true and easy to maintain, it is also a bit underwhelming for IDE users who rely on hovering over a function to see what it does.

For example, if I hover over a call to `fs_err::OpenOptions::create_new()` I just get "Wrapper for OpenOptions::create_new". If I instead hover over `std::fs::OpenOptions::create_new()`, I get a description such as "Sets the option to create a new file, failing if it already exists. [...]". While the shown docs is technically 100% true, it's less useful, and means that swapping `std::fs` for `fs_err` makes the code less convenient to browse in IDE.

Could we make a compromise between maintainability and usefulness by copying the first sentence of the upstream docs? Something like "Sets the option to create a new file, failing if it already exists. Wrapper for OpenOptions::create_new, which see for details."

**Comment by hniksic** [comment]
@andrewhickman Would you accept a PR that implemented such a change?

The downside is that there's no clear process on how to update the docs when the upstream wording changes. But maybe that's acceptable.

**Comment by andrewhickman** [comment]
I'd certainly accept a PR for this.

I don't think the docs getting out of date will be too much of an issue in practise, especially if we only include the first line. It's probably possible to script something using the rustdoc JSON output to check it though

**Comment by schneems** [comment]
> Would you accept a PR that implemented such a change?

In general if you're asking "would you take a PR" and the PR does not involve a new feature (and contributing.md doesn't say otherwise) I find it best to go ahead and send a PR and ask the question there it shortcuts one step in the process and doesn't require blocking on the maintainer (or consuming their time for a simple pre-approval). Especially since many times the work ends up not happening (as in this case, the question was asked in July and still not implemented in September).

For the issue at large I like some kind of script, though a more immediate solution could be to copy all the docs with a note under each.

> This is a wrapper for [fs::read_dir](https://doc.rust-lang.org/stable/std/fs/fn.read_dir.html). If you find the docs have diverged please send a [documentation pr](\<link to github PRs or doc explaining what "doc pr means"\>)

If we wanted to go above and beyond, it would be nice to get an example of what the error message will look like in the docs. If we did this via doc-tests then the suite would catch any upstream changes in wording. 

We could potentially have examples with a link to the official docs in lieu of copying all of the wording (which would get stale faster). This would show a quick "how to use this API at a glance" as well as the extra info of what the error message looks like. If people need more info they can click the "wrapper for" link.

To sum up. I see several paths forward:

- A) Copy upstream docs with a prompt to send a PR for any changes (easy to implement, depending on upstream change velocity might result in maintainer burden to merge and check doc prs).
- B) A script to do that automatically (easier to maintain long term, more difficult for the initial implementation)
- C) Add examples that demonstrate errors to the docs and link to upstream (only needs modification when the upstream error messages change so it's less burdensome to the maintainer than "A", but requires clicking through for the full docs that would be present in "B"

I'm not the maintainer (thanks for your work!) I lean towards option "C" as a pragmatic middle ground.

**Comment by hniksic** [comment]
@andrewhickman 
>I'd certainly accept a PR for this.

One is now open, #50.

**Comment by andrewhickman** [comment]
Released in version 2.11.0

Thanks again for the PR!
